### ISSUE TITLE
no-use-before-define: don't report violation when the definition is in the parent scope of the usage

### ISSUE THREAD

**Issue by camjackson** [eslint/eslint#6659]
Currently using 2.10.2 with eslint-config-airbnb + a few tweaks.

`no-use-before-define` is a great rule, and I definitely want to keep using it, in order to prevent code from relying on hoisting like this:

```js
//bad code
const fivePlusOne = addOne(5);

const addOne = (x) => {
  return x + 1;
}
```

However, to me there's a big difference between the above code, where the usage and definition are in the same scope, and the below code, where the scopes are different.

```js
export const addOne = (x) => {
  return add(x, 1); // I want this to not violate the rule
}

const add = (x, y) => {
  return x + y;
}
```

The usage of `add` in the above code is before the definition in terms of which line they're written on. But in terms of order of execution, the definition happens when the module is loaded, but the usage isn't until the exported function is called. So it's safe regardless of hoisting.

I don't want to turn this rule off, because I want to prevent the kind of hoisting shown in the first example. But to me it's much more natural to define functions in the order that they're called. With the way the rule works at the moment, I'd either need to turn my modules upside down, or ignore the rule.

**Comment by mysticatea** [comment]
Thank you for this issue.

This is duplicate of #6499, #6415, #4533, #4521, #3050, #1485, etc...
This is intentional behavior.

**Comment by camjackson** [comment]
My apologies, I did do a search before posting, but clearly not well enough.

**Comment by alberto** [comment]
Closing as duplicate.
